<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Hypothesis-Testing-in-Python" data-toc-modified-id="Hypothesis-Testing-in-Python-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Hypothesis Testing in Python</a></span></li><li><span><a href="#Two-Sample-Hypothesis-Tests-with-Scipy" data-toc-modified-id="Two-Sample-Hypothesis-Tests-with-Scipy-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Two Sample Hypothesis Tests with Scipy</a></span><ul class="toc-item"><li><span><a href="#Matched-Pairs" data-toc-modified-id="Matched-Pairs-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Matched Pairs</a></span></li><li><span><a href="#Independent-Samples" data-toc-modified-id="Independent-Samples-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Independent Samples</a></span></li></ul></li><li><span><a href="#More-Hypothesis-Testing-with-SciPy-and-Stats-Models" data-toc-modified-id="More-Hypothesis-Testing-with-SciPy-and-Stats-Models-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>More Hypothesis Testing with SciPy and Stats Models</a></span><ul class="toc-item"><li><span><a href="#ANOVA-in-Python" data-toc-modified-id="ANOVA-in-Python-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>ANOVA in Python</a></span></li><li><span><a href="#Linear-Regression" data-toc-modified-id="Linear-Regression-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Linear Regression</a></span></li></ul></li></ul></div>

# Hypothesis Testing in Python

We can use the Scipy library to perform hypothesis tests. The Scipy library has a function for one sample hypothesis tests called ttest_1samp. This test takes a dataset and a constant for comparison and returns the test statistic and the p value for a 2 sided test.

Our test is a one-sided test so we will only look at the test statistic. In order to use the p-value we have to divide the p-value by 2.

In [1]:
import numpy as np
from scipy.stats import ttest_1samp

In [2]:
patients = np.random.normal(5.1, 1.6, 100)
ttest_1samp(patients, 5.7)

Ttest_1sampResult(statistic=-3.000831986203308, pvalue=0.0034069763375635686)

In this example, we generated random data with mean 5.1 and standard deviation 1.6 in order to simulate our patients. Our test statistic is close but not exactly the same since the mean of the sample is not exactly 5.1 like in the example but in fact:

In [3]:
np.mean(patients)

5.240033260244129

This explains the small discrepancy in the test statistic. However, the result is the same - we reject the null hypothesis.

# Two Sample Hypothesis Tests with Scipy

## Matched Pairs

Let's start by looking at our dataset. The file blood_pressure.csv can be downloaded here.

https://s3-eu-west-1.amazonaws.com/ih-materials/uploads/data-static/data/module-2/blood_pressure.csv

In [4]:
import pandas as pd

blood_pressure = pd.read_csv('blood_pressure.csv')
blood_pressure.head()

before       after
0  136.713072   92.432965
1  134.735618  105.022643
2  127.529115   82.242766
3  144.527126   93.607172
4  124.214720  103.212223

We will be using the scipy function ttest_rel. This function is used for hypothesis testing of dependent data.

In [5]:
from scipy.stats import ttest_rel

ttest_rel(blood_pressure.after, blood_pressure.before)

Ttest_relResult(statistic=-27.291841767560236, pvalue=7.303035069608042e-48)

Our result is a very small p-value. This means that we will reject the null hypothesis.

Since a matched pairs test is equivalent to a one sample test of the difference, we can also perform a one sample test and get the exact same result.

In [6]:
from scipy.stats import ttest_1samp
ttest_1samp(blood_pressure.after-blood_pressure.before, 0)

Ttest_1sampResult(statistic=-27.291841767560236, pvalue=7.303035069608042e-48)

We can see that the p-value is identical since the tests are equivalent.

## Independent Samples

In scipy, this means that we will be setting equal_var=True in our function.

The following is an example of a 2 sample hypothesis test with equal variance. We will load a sample dataset of transaction amounts from an e-commerce website. The dataset ab_test.csv can be downloaded here.

https://s3-eu-west-1.amazonaws.com/ih-materials/uploads/data-static/data/module-2/ab_test.csv

In [7]:
ab_test = pd.read_csv('ab_test.csv')
ab_test.head()

a      b
0   0.27  13.61
1   6.08  21.53
2  13.74   9.23
3   9.70   5.36
4   7.00  12.90

We make the assumption that the variances of both populations are equal based on prior knowledge of the data. Now we will test that there is a significant difference between the website layouts with a 95% degree of confidence.

In [8]:
from scipy.stats import ttest_ind

ttest_ind(ab_test.a, ab_test.b, equal_var=True)

Ttest_indResult(statistic=-2.637533181209767, pvalue=0.009713140852447347)

Our p-value is very small. This means that there is a significant difference between the two sample means.

Let's use our A/B test data to perform a t-test that does not require the equal variance assumption:

In [9]:
ttest_ind(ab_test.a, ab_test.b, equal_var=False)

Ttest_indResult(statistic=-2.637533181209767, pvalue=0.009776243024828825)

In this case the p-value slightly differs from the one we get with equal variances. However, since it is very small in this case as well, we will still reject the null hypothesis and conclude that there is a significant difference between the two sample means.

# More Hypothesis Testing with SciPy and Stats Models

## ANOVA in Python

Below is an example of a dataset containing 8 observations of car loan interest rates from 6 different cities. We would like to show that there is a difference in the rates based on city. The dataset rate_by_city.csv can be obtained here.

https://s3-eu-west-1.amazonaws.com/ih-materials/uploads/data-static/data/module-2/rate_by_city.csv

In [10]:
import pandas as pd
from scipy.stats import f_oneway

#let's load the dataset
rate = pd.read_csv('rate_by_city.csv')
rate.head(15)

Rate  City
0   13.75     1
1   13.75     1
2   13.50     1
3   13.50     1
4   13.00     1
5   13.00     1
6   13.00     1
7   12.75     1
8   12.50     1
9   14.25     2
10  13.00     2
11  12.75     2
12  12.50     2
13  12.50     2
14  12.40     2

The dataset contains two columns - rate and city. To test our hypothesis, we need to either pass in multiple filtered subsets to our function or to pivot the dataset to have one column per city. We'll choose the second option. We'll start off by using the cumcount function to create a new index and then use the pivot function to create 6 city columns. We will then rename the columns to allow us to access them more easily.

In [12]:
rate['city_count'] = rate.groupby('City').cumcount()
rate_pivot = rate.pivot(index='city_count', columns='City', values='Rate')
rate_pivot.columns = ['City_'+str(x) for x in rate_pivot.columns.values]
rate_pivot.head()

City_1  City_2  City_3  City_4  City_5  City_6
city_count                                                
0            13.75   14.25   14.00   15.00   14.50   13.50
1            13.75   13.00   14.00   14.00   14.00   12.25
2            13.50   12.75   13.51   13.75   14.00   12.25
3            13.50   12.50   13.50   13.59   13.90   12.00
4            13.00   12.50   13.50   13.25   13.75   12.00

Now that we have successfully pivoted the data, we can perform the test. The f_oneway function requires us to specify each column that is passed into the function (rather than passing the entire dataframe)

In [13]:
f_oneway(rate_pivot.City_1,rate_pivot.City_2,rate_pivot.City_3,rate_pivot.City_4,rate_pivot.City_5,rate_pivot.City_6)

F_onewayResult(statistic=4.8293848737024, pvalue=0.001174551414504048)

The p-value is 0.001174. This value is very small, certainly smaller than 0.05. Therefore, we reject the null hypothesis and conclude that the rates differ by city.

The function for generating an ANOVA in statsmodels is called anova_lm and it generates an ANOVA table. As a first step, we define a model and then generate the ANOVA table. In this case, we prefer not to pivot our data since the library will do it for us.

In [14]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

model = ols('Rate ~ C(City)', data=rate).fit()
anova_table = sm.stats.anova_lm(model, typ=2)
anova_table


#sum_sqfloat64
#Sum of squares for model terms.

#dffloat64
#Degrees of freedom for model terms.

#Ffloat64
#F statistic value for significance of adding model terms.

#PR(>F)float64
#P-value for significance of adding model terms.

sum_sq    df         F    PR(>F)
C(City)   10.945667   5.0  4.829385  0.001175
Residual  21.758133  48.0       NaN       NaN

In the code below, we defined a model of rate and city. The pivoting is performed internally by using the C function. Our result is the same p-value and our conclusion to reject remains the same.

## Linear Regression

In the example below, we will create a linear model that predicts MPG using acceleration in the auto-mpg dataset. Note that linregress only supports linear regression with one variable for x and one for y.

The dataset auto-mpg.csv can be obtained here.

https://s3-eu-west-1.amazonaws.com/ih-materials/uploads/data-static/data/auto-mpg.csv

In [15]:
from scipy.stats import linregress

auto = pd.read_csv('auto-mpg.csv')
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

In [16]:
#Calculate a linear least-squares regression for two sets of measurements.

slope, intercept, r_value, p_value, std_err = linregress(auto.acceleration, auto.mpg)

In [17]:
slope, intercept, r_value, p_value, std_err

#slopefloat
#Slope of the regression line.

#interceptfloat
#Intercept of the regression line.

#rvaluefloat
#Correlation coefficient.

#pvaluefloat
#Two-sided p-value for a hypothesis test whose null hypothesis is that the slope is zero, using Wald Test with t-distribution of the test statistic.

#stderrfloat
#Standard error of the estimated gradient.

(1.1912045293502274,
 4.9697930042539085,
 0.4202889121016507,
 1.8230915350787203e-18,
 0.12923643283101396)

This means that our regression equation is:

mpg = 4.9698 + 1.1912 * acceleration

The r squared is 0.1766 which is relatively small. This means that our model only captures 17% of the variation in the data.

The p-value is very small, this means that the slope is significantly different from zero.

In [18]:
import statsmodels.api as sm

X = sm.add_constant(auto.acceleration) # We must add the intercept using the add_constant function
Y = auto.mpg

model = sm.OLS(Y, X).fit()
predictions = model.predict(X) 

print_model = model.summary()
print(print_model)


#El criterio de información de Akaike (AIC) es una medida de la calidad relativa de un modelo estadístico, para un conjunto dado de datos. Como tal, el AIC proporciona un medio para la selección del modelo.
#AIC maneja un trade-off entre la bondad de ajuste del modelo y la complejidad del modelo.

#The Bayesian information criterion (BIC) or Schwarz information criterion (also SIC, SBC, SBIC) is a criterion for model selection among a finite set of models; the model with the lowest BIC is preferred. It is based, in part, on the likelihood function and it is closely related to the Akaike information criterion (AIC).

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.177
Model:                            OLS   Adj. R-squared:                  0.175
Method:                 Least Squares   F-statistic:                     84.96
Date:                Mon, 05 Aug 2019   Prob (F-statistic):           1.82e-18
Time:                        08:43:08   Log-Likelihood:                -1343.9
No. Observations:                 398   AIC:                             2692.
Df Residuals:                     396   BIC:                             2700.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            4.9698      2.043      2.432   

//anaconda3/envs/Bismart_Raquel/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


Here, we are not limited to only one predictor variable. Let's try this regression with more than one predictor.

In [19]:
X = sm.add_constant(auto[['cylinders', 'weight', 'acceleration']]) # adding a constant
Y = auto.mpg

model = sm.OLS(Y, X).fit()
predictions = model.predict(X) 

print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.700
Model:                            OLS   Adj. R-squared:                  0.698
Method:                 Least Squares   F-statistic:                     306.7
Date:                Mon, 05 Aug 2019   Prob (F-statistic):          1.14e-102
Time:                        08:43:40   Log-Likelihood:                -1142.9
No. Observations:                 398   AIC:                             2294.
Df Residuals:                     394   BIC:                             2310.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const           42.3811      1.960     21.627   